src/main/java/io/vertx/mqtt/impl/MqttClientSessionImpl.java

```diff
@@ -0,0 +1,578 @@
+      if (log.isDebugEnabled()) {
+        log.debug("Timer set: " + timer);
+      }
+
+      this.reconnectTimer = timer;
+    }
+  }
+
+  /**
+   * Calculate the next delay before trying to re-connect.
+   *
+   * @return The duration to wait.
+   */
+  private Duration nextDelay() {
+    return Duration.ofSeconds(10);
+  }
+
+  /**
+   * Initiates the connection.
+   */
+  private void createConnection() {
+    log.debug("Creating connection");
+
+    // clear reconnect timer
+    this.reconnectTimer = null;
+
+    // create client
+    this.client = MqttClient.create(this.vertx, this.options);
+    this.client.exceptionHandler(this::closeConnection);
+    this.client.closeHandler(x -> connectionClosed());
+    this.client.publishHandler(this::serverPublished);
+    this.client.subscribeCompletionHandler(this::subscribeCompleted);
+    this.client.unsubscribeCompletionHandler(this::unsubscribeCompleted);
+    this.client.publishCompletionHandler(this::publishComplete);
+
+    // change state
+    setState(SessionState.CONNECTING, null);
+    // start connection
+    this.client.connect(this.options.getPort(), this.options.getHostname()).onComplete(this::connectCompleted);
+  }
+
+  /**
+   * Initiates the connection shutdown.
+   */
+  private void closeConnection(Throwable cause) {
+    log.debug("Closing connection", cause);
+
+    setState(SessionState.DISCONNECTING, cause);
+    this.client.disconnect().onComplete(this::disconnectCompleted);
+  }
+
+  /**
+   * Gets called when the connect call was processed.
+   *
+   * @param result The outcome of the connect call.
+   */
+  private void connectCompleted(AsyncResult<MqttConnAckMessage> result) {
+
+    if (log.isDebugEnabled()) {
+      log.debug(String.format("Connect completed - result: %s, cause: %s", result.result(), result.cause()));
+    }
+
+    if (result.failed() || result.result() == null) {
+      // this will include CONACKs with error codes
+      setState(SessionState.DISCONNECTED, result.cause());
+      return;
+    }
+
+    setState(SessionState.CONNECTED, null);
+
+    if (!this.subscriptions.isEmpty()) {
+      requestSubscribe(new LinkedHashMap<>(this.subscriptions));
+    }
+  }
+
+  /**
+   * Gets called when the disconnect call was processed.
+   *
+   * @param result The outcome of the disconnect call.
+   */
+  private void disconnectCompleted(AsyncResult<?> result) {
+
+    if (log.isDebugEnabled()) {
+      log.debug(String.format("Disconnect completed - result: %s, cause: %s", result.result(), result.cause()));
+    }
+
+    connectionClosed(result.cause());
+  }
+
+  /**
+   * Gets called internally when the only reasonable action is to just disconnect.
+   * <p>
+   * If the session is still running, then it will trigger a re-connect.
+   *
+   * @param reason The reason message.
+   */
+  private void closeConnection(final String reason) {
+    closeConnection(new IOException(reason));
```

Comment:
might be better to use a VertxEception, as anyway the stack trace is useless.

Reply:
Done